src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java

```diff
@@ -325,13 +332,34 @@ private static synchronized ScheduledThreadPoolExecutor threadPool() {
             return false;
         }
 
-        private @CheckForNull FilePath getWorkspace() throws AbortException {
+        private @CheckForNull FilePath getWorkspace() throws IOException, InterruptedException {
             if (ws == null) {
                 ws = FilePathUtils.find(node, remote);
                 if (ws == null) {
+                    // Part of JENKINS-49707: check whether an agent has been removed.
+                    // (Note that a Computer may be missing because a Node is offline,
+                    // and conversely after removing a Node its Computer may remain for a while.
+                    // Therefore we only fail here if _both_ are absent.)
+                    // ExecutorStepExecution.NodeListener will normally do this first, so this is a fallback.
+                    Jenkins j = Jenkins.getInstanceOrNull();
+                    if (ExecutorStepExecution.RemovedNodeCause.ENABLED && !node.isEmpty() && j != null && j.getNode(node) == null) {
+                        if (removedNodeDiscovered == 0) {
+                            LOGGER.fine(() -> "discovered that " + node + " has been removed");
+                            removedNodeDiscovered = System.nanoTime();
+                            return null;
+                        } else if (System.nanoTime() < removedNodeDiscovered + TimeUnit.MILLISECONDS.toNanos(ExecutorPickle.TIMEOUT_WAITING_FOR_NODE_MILLIS)) {
```

Comment:
Do we need to switch to the following as in the `System.nanoTime()` JavaDoc to avoid potential overflow issues (or just switch to `System.currentTimeMillis()` if nanosecond precision is not important)?

```suggestion
                        } else if (System.nanoTime() - removedNodeDiscovered < TimeUnit.MILLISECONDS.toNanos(ExecutorPickle.TIMEOUT_WAITING_FOR_NODE_MILLIS)) {
```

Reply:
`currentTimeMillis` is inappropriate for deltas since it could drift with clock changes like NTP does.

The overflow issue only matters if Jenkins stays running for at least 292 years without a restart but OK!

Reply:
> currentTimeMillis is inappropriate for deltas since it could drift with clock changes like NTP does

Ah I didn't realize `System.nanoTime` was divorced from real-world time and only intended for elapsed time (a bit of selective Javadoc reading on my part).